core/handler.go

```diff
@@ -0,0 +1,49 @@
+package core
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/filecoin-project/go-filecoin/chain"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+// NewHeadHandler wires up new head tipset handling to the message inbox and outbox.
+type NewHeadHandler struct {
+	// Inbox and outbox exported for testing.
+	Inbox  *Inbox
+	Outbox *Outbox
+	chain  outboxChainProvider
+
+	prevHead types.TipSet
+}
+
+// NewHandler build a new new-head handler.
+func NewHandler(inbox *Inbox, outbox *Outbox, chain outboxChainProvider, head types.TipSet) *NewHeadHandler {
+	return &NewHeadHandler{inbox, outbox, chain, head}
+}
+
+// HandleNewHead computes the chain delta implied by a new head and updates the inbox and outbox.
+func (h *NewHeadHandler) HandleNewHead(ctx context.Context, newHead types.TipSet) error {
+	if !newHead.Defined() {
+		return errors.New("received empty tipset, ignoring")
+	}
+	if newHead.Equals(h.prevHead) {
+		return errors.Errorf("received non-new head tipset, ignoring %s", newHead.Key())
```

Comment:
Are these both really errors? I could see the first one but maybe this one isn't.  If it's determined the head isn't new at all it seems to me it's still handled -- it could log this message then return nil.